django/db/models/sql/query.py

```diff
@@ -1142,6 +1143,10 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
         """
         if isinstance(filter_expr, dict):
             raise FieldError("Cannot parse keyword query as dict")
+        if isinstance(filter_expr, Expression) and isinstance(filter_expr.output_field, BooleanField):
```

Comment:
It's only in two places and I don't think it cleans up that much, you'd just end up with `if isinstance(filter_expr, Expression) and filter_expr.is_output_field_boolean():` which is slightly less explicit and slightly slower